[lldb] Re-use clang's keyword enable/disable mechanism in CPlusPlusNameParser.cpp

dsandersllvm opened the pull request:
The problem is that CPlusPlusNameParser doesn't respect the language dialect for `target variable Class::constant`. If `constant` is a keyword in some (downstream in this case) dialects then it lexes `constant` as `kw_constant` instead of `kw_raw_identifier` even if that dialect is not active. As a result:
* `target variable constant` works
* `target variable Class::constant` fails to look up constant because it sees `kw_raw_identifier, kw_coloncolon, kw_constant` instead of `kw_raw_identifier, kw_coloncolon, kw_raw_identifier`
* `expr -l c++ -- Class::constant` works because clang is parsing it

Note: There's seemingly a separate bug where GetLangOptions() does not account for the process/frame language and just uses a pre-defined set of language options. I have not attempted to fix that since, for now, at least hardcoding my dialect to be disabled by that function does the right thing for existing tests.

Also fixed a trivial return-after-else that clangd pointed out in the same area

dsandersllvm committed:
[clang] Refactor clang's keyword enable/disable mechanism to allow lldb to re-use it

lldb's CPlusPlusNameParser is currently identifying keywords using it's own
map implemented using clang/Basic/TokenKinds.def. However, it does not respect
the language options so identifiers can incorrectly determined to be keywords
when using languages in which they are not keywords.

Rather than implement the logic to enable/disable keywords in both projects it
makes sense for lldb to use clang's implementation.

See #164284 for more information

dsandersllvm committed:
[clang] Refactor clang's keyword enable/disable mechanism to allow lldb to re-use it

lldb's CPlusPlusNameParser is currently identifying keywords using it's own
map implemented using clang/Basic/TokenKinds.def. However, it does not respect
the language options so identifiers can incorrectly determined to be keywords
when using languages in which they are not keywords.

Rather than implement the logic to enable/disable keywords in both projects it
makes sense for lldb to use clang's implementation.

See #164284 for more information

dsandersllvm committed:
[lldb] Re-use clang's keyword enable/disable mechanism in CPlusPlusNameParser.cpp

The problem is that CPlusPlusNameParser doesn't respect the language dialect
for `target variable Class::constant`. If `constant` is a keyword in some
(downstream in this case) dialects then it lexes `constant` as `kw_constant`
instead of `kw_raw_identifier` even if that dialect is not active. As a
result:
* `target variable constant` works
* `target variable Class::constant` fails to look up constant because it sees
  `kw_raw_identifier, kw_coloncolon, kw_constant` instead of
  `kw_raw_identifier, kw_coloncolon, kw_raw_identifier`
* `expr -l c++ -- Class::constant` works because clang is parsing it

Note: There's seemingly a separate bug where GetLangOptions() does not account
for the process/frame language and just uses a pre-defined set of language
options. I have not attempted to fix that since, for now, at least
hardcoding my dialect to be disabled by that function does the
right thing for existing tests.

Also fixed a trivial return-after-else that clangd pointed out in the same
area

dsandersllvm committed:
fixup: use `using namespace clang` instead

dsandersllvm committed:
Revert "[lldb] Re-use clang's keyword enable/disable mechanism in CPlusPlusNameParser.cpp"

This is just to allow me to separate the clang and lldb parts without breaking
the github PR's comment history

dsandersllvm committed:
[lldb] Re-use clang's keyword enable/disable mechanism in CPlusPlusNameParser.cpp

The problem is that CPlusPlusNameParser doesn't respect the language dialect
for `target variable Class::constant`. If `constant` is a keyword in some
(downstream in this case) dialects then it lexes `constant` as `kw_constant`
instead of `kw_raw_identifier` even if that dialect is not active. As a
result:
* `target variable constant` works
* `target variable Class::constant` fails to look up constant because it sees
  `kw_raw_identifier, kw_coloncolon, kw_constant` instead of
  `kw_raw_identifier, kw_coloncolon, kw_raw_identifier`
* `expr -l c++ -- Class::constant` works because clang is parsing it

Note: There's seemingly a separate bug where GetLangOptions() does not account
for the process/frame language and just uses a pre-defined set of language
options. I have not attempted to fix that since, for now, at least
hardcoding my dialect to be disabled by that function does the
right thing for existing tests.

Also fixed a trivial return-after-else that clangd pointed out in the same
area

dsandersllvm committed:
fixup: formatting

Michael137 reviewed (commented):
`lldb/unittests/Language/CPlusPlus/CPlusPlusLanguageTest.cpp` has some unit-tests for `CPlusPlusNameParser` (and friends). If we added a constructor to `CPlusPlusNameParser` which took a user-provided `LangOptions` (but defaults to the existing one) that we could add a unit-test which constructs the `CPlusPlusNameParser` with some `LangOptions` which triggers the issue you were seeing. What do you think?

dsandersllvm committed:
fixup: Move getKeywordStatus's comment to the header

dsandersllvm committed:
fixup: Restore comment for TokenKey

This comment was originally on the namespace containing TokenKey and
KeywordStatus. I've only restored it to TokenKey because KeywordStatus
is used by the code that uses TokenKinds.def not TokenKinds.def itself.

dsandersllvm committed:
Merge branch 'lldb-parse-identifiers-keywords-clang-only' into lldb-parse-identifiers-keywords

dsandersllvm committed:
fixup: Move using

dsandersllvm committed:
fixup: Test Class::constant in CPlusPlusLanguageTest.cpp

dsandersllvm commented on lldb/source/Plugins/Language/CPlusPlus/CPlusPlusNameParser.cpp:
Done
